### ISSUE TITLE
[yul] datasize use can be problematic if compiler generated code uses it in a stateful way

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#11035]
## Description

```
object "object0" {
  code {
    {
      mstore(0, 0)
      sstore(add(datasize("object0"),0), 0x80000000000000000000000000000000000000)
    }
  }
}
```

will lead to different storage pre/post optimisation. This is expected because the `mstore(0,0)` is going to be optimised out by the optimiser (haven't verified this but I'm pretty sure it will be), shrinking the object such that `datasize` for the optimised case is going to be off by a byte or so (and hence execute the store at a different slot).

Related to #7411

**Comment by hrkrshnn** [comment]
I don't think `mstore(0, 0)` will be optimized out. Currently we never remove unused writes to memory.

**Comment by hrkrshnn** [comment]
Although, you could probably replace `mstore` with something else that would be removed, right? Currently, we don't seem to consider the side effect of `datasize` anywhere. Not sure if we should.

**Comment by bshastry** [comment]
> I don't think `mstore(0, 0)` will be optimized out. Currently we never remove unused writes to memory.

Ah sorry, I somehow thought we did. I think what's happening is that the `add(datasize(..), 0` is optimized to just `datasize(..)` in that case leading to a couple of fewer bytes (not sure exactly but I guess 1 additional byte for add, and 2 bytes for PUSH0 and 0x0)

**Comment by bshastry** [comment]
> Currently, we don't seem to consider the side effect of `datasize` anywhere. Not sure if we should.

Me too. I'm trying to think one can make a different argument than what we did for `msize` (see #7411). It may be unlikely that the compiler generated code would contain `datasize` in the way that would lead to a side-effect in which case the only circumstance in which this would be problematic is user-written code but with the latter I doubt if there is a real use-case.

**Comment by hrkrshnn** [comment]
About `msize`: I would say it's an exception since a user can always use it in inline assembly. I don't think `solc` generates `msize` otherwise. Currently, many steps are disabled if the AST has `msize`.

The situation here is akin to the following: `extcodesize(<address>)` would have different value depending on whether the contract to be deployed at `<address>` is optimized or not. 

I think we can ignore this problem. Would this lead to a lot of false positives?

----
Also, the self-referential `datasize` doesn't look right. Let me see if that is well-defined.

**Comment by bshastry** [comment]
> I think we can ignore this problem. Would this lead to a lot of false positives?

I agree. Since this issue only occurs when driving evmone and not the yul interpreter (datasize in yul interpreter returns a constant), this is somewhat particular to https://github.com/ethereum/solidity/pull/11028 that is in draft mode.

I think we can work around by simply not generating "problematic" uninteresting opcodes from a bug-finding point of view like address(), gas(), datasize() etc.

May be implemented using a flag (meant to be used when targeting evm client).

**Comment by bshastry** [comment]
This may be closed once #11028 is merged because this ~~bug~~ issue only appears while fuzzing on evmone.
